runners/direct-java/src/main/java/org/apache/beam/runners/direct/WriteWithShardingFactory.java

```diff
@@ -57,92 +49,51 @@
+  public PTransform<PCollection<InputT>, PDone> getReplacementTransform(Bound<InputT> transform) {
     if (transform.getSharding() == null) {
-      return new DynamicallyReshardedWrite<>(transform);
+      return transform.withSharding(new LogElementShardsWithDrift<InputT>());
     }
     return transform;
   }
 
   @Override
-  public PCollection<InputT> getInput(
-      List<TaggedPValue> inputs, Pipeline p) {
+  public PCollection<InputT> getInput(List<TaggedPValue> inputs, Pipeline p) {
     return (PCollection<InputT>) Iterables.getOnlyElement(inputs).getValue();
   }
 
   @Override
-  public Map<PValue, ReplacementOutput> mapOutputs(
-      List<TaggedPValue> outputs, PDone newOutput) {
+  public Map<PValue, ReplacementOutput> mapOutputs(List<TaggedPValue> outputs, PDone newOutput) {
     return Collections.emptyMap();
   }
 
-  private static class DynamicallyReshardedWrite<T> extends PTransform<PCollection<T>, PDone> {
-    private final transient Write.Bound<T> original;
-
-    private DynamicallyReshardedWrite(Bound<T> original) {
-      this.original = original;
-    }
+  private static class LogElementShardsWithDrift<T>
+      extends PTransform<PCollection<T>, PCollectionView<Integer>> {
 
     @Override
-    public PDone expand(PCollection<T> input) {
-      checkArgument(
-          IsBounded.BOUNDED == input.isBounded(),
-          "%s can only be applied to a Bounded PCollection",
-          getClass().getSimpleName());
-      PCollection<T> records =
-          input.apply(
-              "RewindowInputs",
-              Window.<T>into(new GlobalWindows())
-                  .triggering(DefaultTrigger.of())
-                  .withAllowedLateness(Duration.ZERO)
-                  .discardingFiredPanes());
-      final PCollectionView<Long> numRecords =
-          records.apply("CountRecords", Count.<T>globally().asSingletonView());
-      PCollection<T> resharded =
-          records
-              .apply(
-                  "ApplySharding",
-                  ParDo.withSideInputs(numRecords)
-                      .of(
-                          new KeyBasedOnCountFn<T>(
-                              numRecords,
-                              ThreadLocalRandom.current().nextInt(MAX_RANDOM_EXTRA_SHARDS))))
-              .apply("GroupIntoShards", GroupByKey.<Integer, T>create())
-              .apply("DropShardingKeys", Values.<Iterable<T>>create())
-              .apply("FlattenShardIterables", Flatten.<T>iterables());
-      // This is an inverted application to apply the expansion of the original Write PTransform
-      // without adding a new Write Transform Node, which would be overwritten the same way, leading
-      // to an infinite recursion. We cannot modify the number of shards, because that is determined
-      // at runtime.
-      return resharded.apply(original);
+    public PCollectionView<Integer> expand(PCollection<T> records) {
+      final int randomExtraShards = ThreadLocalRandom.current().nextInt(0, 3);
```

Comment:
This can actually be inside the `CalculateShardsDoFn`, as it's only called once per-window and put in a `SideInput`, so I've moved it there.